lib/pbench/cli/server/report.py

```diff
@@ -0,0 +1,403 @@
+    click.echo("SQL storage report:")
+    click.echo("  Table                Rows       Storage")
+    click.echo("  -------------------- ---------- ----------")
+    for t in inspect(Database.db_session.get_bind()).get_table_names():
+        rows = list(
+            Database.db_session.execute(statement=text(f"SELECT COUNT(*) FROM {t}"))
+        )[0][0]
+        size = list(
+            Database.db_session.execute(
+                statement=text("SELECT pg_total_relation_size(:table)"),
+                params={"table": t},
+            )
+        )[0][0]
+        click.echo(f"  {t:20} {rows:>10} {humanize.naturalsize(size):>10}")
+
+    if not detailer:
+        return
+
+    query = select(IndexMap.root, IndexMap.index)
+    idxes = Database.db_session.execute(query).all()
+    record_count = 0
+    roots = set()
+    indices = set()
+    root_size = 0
+    index_size = 0
+    for idx in idxes:
+        record_count += 1
+        roots.add(idx[0])
+        indices.add(idx[1])
+        root_size += len(idx[0])
+        index_size += len(idx[1])
+    unique_root_size = 0
+    unique_index_size = 0
+    for r in roots:
+        unique_root_size += len(r)
+    for i in indices:
+        unique_index_size += len(i)
+
+    detailer.write(
+        f"{record_count} indexmap records found with {len(indices)} indices "
+        f"and {len(roots)} roots:"
+    )
+    detailer.write(
+        f" {humanize.naturalsize(index_size)} for index names, "
+        f"{humanize.naturalsize(root_size)} for root names"
+    )
+    detailer.write(
+        f" deduped: {humanize.naturalsize(unique_index_size)} for index "
+        f"names, {humanize.naturalsize(unique_root_size)} for root names"
+    )
+
+
+def report_states():
+    """Report tarball operational states."""
+
+    index_pattern: re.Pattern = re.compile(r"^(\d+):(.*)$")
+    index_errors = defaultdict(int)
+    index_messages = defaultdict(str)
+    operations = defaultdict(lambda: defaultdict(int))
+    rows = Database.db_session.execute(
+        statement=text(
+            "SELECT d.name, o.name, o.state, o.message FROM datasets AS d LEFT OUTER JOIN "
+            "dataset_operations AS o ON o.dataset_ref = d.id"
+        )
+    )
+    for row in rows:
+        operations[row[1]][row[2]] += 1
+        if row[2] == "FAILED":
+            detailer.write(f"{row[1]} {row[2]} {row[0]} {row[3]!r}")
+            if row[1] == "INDEX":
+                match = index_pattern.match(row[3])
+                if match:
+                    try:
+                        code = int(match.group(1))
+                        message = match.group(2)
+                        index_errors[code] += 1
+                        if code not in index_messages:
+                            index_messages[code] = message
+                    except Exception as e:
+                        verifier.status(
+                            f"{row[0]} unexpected 'INDEX' error {row[3]}: {str(e)!r}"
+                        )
```

Comment:
I don't think you need the `try` block:  if the regex matches, then what could trigger an exception?

Reply:
I was clearly being paranoid, as the message indicates. Whether it was unreasonably paranoid, or remains so, are different questions. 😆

Reply:
So, you've decided that the paranoia is reasonable (and I should resolve this conversation)?

Reply:
I'm not dissatisfied with my level of paranoia. 😆